docs/spec/relayer/Packets.md

```diff
@@ -0,0 +1,267 @@
+at the most recent height. 
+
+
+## Computing destination chain
+
+```golang
+func GetDestinationInfo(ev IBCEvent, chainA Chain) Chain {
+    switch ev.type {
+        case SendPacketEvent: 
+            channel, proof = GetChannel(chain, ev.sourcePort, ev.sourceChannel, ev.Height)
+            if proof == nil return nil
+                
+            connectionId = channel.connectionHops[0]
+            connection, proof = GetConnection(chain, connectionId, ev.Height) 
+            if proof == nil return nil
+                
+            clientState = GetClientState(chain, connection.clientIdentifier, ev.Height) 
+            return getHostInfo(clientState.chainID) 
+        ...    
+    }
+}
+```
+
+## Datagram creation logic
+
+### PacketRecv datagram creation
+
+```golang
+func createPacketRecvDatagram(ev SendPacketEvent, chainA Chain, chainB Chain, installedHeight Height) PacketRecv {        
+    
+    // Stage 1 
+    // Verify if packet is committed to chain A and it is still pending (commitment exists)
+    
+    proofHeight = installedHeight - 1
+    packetCommitment, packetCommitmentProof = 
+        GetPacketCommitment(chainA, ev.sourcePort, ev.sourceChannel, ev.sequence, proofHeight)     
+    if packetCommitmentProof != nil { return nil }
+        
+    if packetCommitment == null OR
+       packetCommitment != hash(concat(ev.data, ev.timeoutHeight, ev.timeoutTimestamp)) { return nil }
+            
+    // Stage 2 
+    // Execute checks IBC handler on chainB will execute
+    
+    channel, proof = GetChannel(chainB, ev.destPort, ev.destChannel, LATEST_HEIGHT)
+    if proof != nil { return nil }
+    
+    if channel == null OR
+       channel.state != OPEN OR
+       ev.sourcePort != channel.counterpartyPortIdentifier OR
+       ev.sourceChannel != channel.counterpartyChannelIdentifier { return nil }
+    
+    connectionId = channel.connectionHops[0]
+    connection, proof = GetConnection(chainB, connectionId, LATEST_HEIGHT) 
+    if proof != nil { return nil }
+    
+    if connection == null OR connection.state != OPEN { return nil } 
+    
+    if ev.timeoutHeight != 0 AND GetConsensusHeight(chainB) >= ev.timeoutHeight { return nil }
+    if ev.timeoutTimestamp != 0 AND GetCurrentTimestamp(chainB) >= ev.timeoutTimestamp { return nil }
+    
+    // we now check if this packet is already received by the destination chain
+    if (channel.ordering === ORDERED) {    
+        nextSequenceRecv, proof = GetNextSequenceRecv(chainB, ev.destPort, ev.destChannel, LATEST_HEIGHT) 
+        if proof != nil { return nil }
+        
+        if ev.sequence != nextSequenceRecv { return nil } // packet has already been delivered by another relayer
+    
+    } else {
+        packetAcknowledgement, proof = 
+            GetPacketAcknowledgement(chainB, ev.destPort, ev.destChannel, ev.sequence, LATEST_HEIGHT)
+        if proof != nil { return nil }
+
+        if packetAcknowledgement != nil { return nil }
```

Comment:
I believe the proof returned by `GetPacketAcknowledgment` is a proof of non-membership if `packetAcknowledgment` is `nil` and we should be able to verify that.

Reply:
Very good point; I will double check this.